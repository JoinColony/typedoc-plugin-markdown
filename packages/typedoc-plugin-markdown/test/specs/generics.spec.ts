import * as Handlebars from 'handlebars/runtime';
import { ProjectReflection, SignatureReflection } from 'typedoc';

describe(`Generics:`, () => {
  let project: ProjectReflection;
  let partial: Handlebars.TemplateDelegate;
  let declarationPartial: Handlebars.TemplateDelegate;
  let reflectionTemplate: Handlebars.TemplateDelegate;

  beforeAll(async () => {
    const bootstrap = global.bootstrap('generics.ts', {
      stubPartials: [
        'comment',
        'members',
        'member.signature',
        'member.sources',
      ],
      stubHelpers: ['toc', 'breadcrumbs', 'hierarchy', 'returns'],
    });
    project = bootstrap.project;
    partial = global.getTemplate(bootstrap.context, 'member.signature');
    declarationPartial = global.getTemplate(
      bootstrap.context,
      'member.declaration',
    );
    reflectionTemplate = global.getTemplate(
      bootstrap.context,
      'reflection',
      false,
    );
  });

  test(`should compile class with type params`, () => {
    expect(
      global.renderTemplate(reflectionTemplate, {
        model: project.findReflectionByName('ClassWithTypeParams'),
        project: project,
      }),
    ).toMatchSnapshot();
  });

  test(`should compile function with a simple type param'`, () => {
    expect(
      global.renderTemplate(
        partial,
        (project.findReflectionByName('functionWithTypeParam') as any)
          .signatures[0] as SignatureReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile function with complex type params'`, () => {
    expect(
      global.renderTemplate(
        partial,
        (project.findReflectionByName('functionWithTypeParams') as any)
          .signatures[0] as SignatureReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile type with nested generics'`, () => {
    expect(
      global.renderTemplate(
        declarationPartial,
        project.findReflectionByName('nestedGenerics'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile generics with defaults'`, () => {
    expect(
      global.renderTemplate(
        declarationPartial,
        (project.findReflectionByName('genericsWithDefaults') as any)
          .signatures[0] as SignatureReflection,
      ),
    ).toMatchSnapshot();
  });
});
