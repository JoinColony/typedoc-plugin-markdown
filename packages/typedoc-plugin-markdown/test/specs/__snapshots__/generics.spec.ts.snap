// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generics: should compile class with type params 1`] = `
"[breadcrumbs]

# Class: ClassWithTypeParams<T, V\\\\>

Comments for typeParams

## Type parameters

| Name | Description |
| :------ | :------ |
| \`T\` | Some type param |
| \`V\` | Some other type param |

[helper: toc]

[groups]
"
`;

exports[`Generics: should compile function with a simple type param' 1`] = `
"▸ **functionWithTypeParam**<\`A\`\\\\>(): \`boolean\`

#### Type parameters

| Name |
| :------ |
| \`A\` |

#### Returns

\`boolean\`

[sources]
"
`;

exports[`Generics: should compile function with complex type params' 1`] = `
"▸ **functionWithTypeParams**<\`A\`, \`B\`, \`C\`\\\\>(): \`boolean\`

#### Type parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| \`A\` | extends [\`ClassWithTypeParams\`](ClassWithTypeParams.md)<\`string\`, \`number\`, \`A\`\\\\> | Comment for type \`A\` |
| \`B\` | \`string\` \\\\| \`boolean\` | Comment for type \`B\` |
| \`C\` | \`string\` | - |

#### Returns

\`boolean\`

[sources]
"
`;

exports[`Generics: should compile generics with defaults' 1`] = `
"• **genericsWithDefaults**()

[sources]
"
`;

exports[`Generics: should compile type with nested generics' 1`] = `
"Ƭ **nestedGenerics**(): [\`Generic1\`](../modules.md#generic1)<[\`Generic2\`](../modules.md#generic2)<[\`Generic3\`](../modules.md#generic3)<\`string\`\\\\>\\\\>\\\\>

#### Returns

[\`Generic1\`](../modules.md#generic1)<[\`Generic2\`](../modules.md#generic2)<[\`Generic3\`](../modules.md#generic3)<\`string\`\\\\>\\\\>\\\\>

[sources]
"
`;
